api/admin/controller/__init__.py

```diff
@@ -2483,3 +2501,23 @@ def _is_email(self, email):
         """Email addresses must be in the format 'x@y.z'."""
         email_format = ".+\@.+\..+"
         return re.search(email_format, email)
+
+    def validate_url(self, settings):
+        """Find any URLs that the user has submitted, and make sure that
+        they are in a valid format."""
+        settings = settings or self._get_settings()
+        if isinstance(settings, (list,)):
+            # Find the fields that have to do with URLs
+            url_fields = filter(lambda s: (s.get("format") == "url" or s.get("key") == "url") and flask.request.form.get(s.get("key")), settings)
+            # Narrow the URL-related fields down to the ones for which the user actually entered a value
+            url_inputs = [flask.request.form.get(field.get("key")) for field in url_fields]
+        else:
+            # The SitewideSettingsController passes in a string
+            url_inputs = [settings]
+        for url in url_inputs:
+            if not self._is_url(url):
+                return INVALID_URL.detailed(_('"%(url)s" is not a valid URL.', url=url))
+
+    def _is_url(self, url):
+        url_format = "(localhost|http[s]?://)(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+"
```

Comment:
I'm not sure about the `localhost` thing; we'll take a look once I go through the rest.